robot/impl/resource_manager_test.go

```diff
@@ -1583,6 +1584,188 @@ func TestReconfigure(t *testing.T) {
 	}()
 }
 
+func TestRemoteConnClosedOnReconfigure(t *testing.T) {
+	logger, observer := logging.NewObservedTestLogger(t)
+
+	ctx := context.Background()
+
+	fakeMotor := resource.Config{
+		Name:                "motor",
+		Model:               resource.DefaultModelFamily.WithModel("fake"),
+		API:                 motor.API,
+		ConvertedAttributes: &fakemotor.Config{},
+	}
+
+	fakeArm := resource.Config{
+		Name:                "arm",
+		Model:               resource.DefaultModelFamily.WithModel("fake"),
+		API:                 arm.API,
+		ConvertedAttributes: &fakearm.Config{},
+	}
+
+	remoteCfg1 := &config.Config{
+		Components: []resource.Config{fakeMotor},
+	}
+
+	remoteCfg2 := &config.Config{
+		Components: []resource.Config{fakeArm},
+	}
+
+	t.Run("remotes with same exact resources", func(t *testing.T) {
+		remote1 := setupLocalRobot(t, ctx, remoteCfg1, logger.Sublogger("remote1"))
+		remote2 := setupLocalRobot(t, ctx, remoteCfg1, logger.Sublogger("remote2"))
+
+		options1, _, addr1 := robottestutils.CreateBaseOptionsAndListener(t)
+		err := remote1.StartWeb(ctx, options1)
+		test.That(t, err, test.ShouldBeNil)
+
+		options2, _, addr2 := robottestutils.CreateBaseOptionsAndListener(t)
+		err = remote2.StartWeb(ctx, options2)
+		test.That(t, err, test.ShouldBeNil)
+
+		// Set up a local main robot which is connected to remote1
+		remoteConf := config.Remote{
+			Name:    "remote",
+			Address: addr1,
+		}
+
+		mainRobotCfg := &config.Config{
+			Remotes: []config.Remote{remoteConf},
+		}
+
+		// Make a copy of the main robot config as reconfigure will directly modify it
+		mainCfgCopy := *mainRobotCfg
+		mainRobot := setupLocalRobot(t, ctx, mainRobotCfg, logger.Sublogger("main"))
+
+		// Grab motor of remote1 to check that it won't work after switching remotes
+		motor1, err := motor.FromRobot(mainRobot, "motor")
+		test.That(t, err, test.ShouldBeNil)
+
+		moving, speed, err := motor1.IsPowered(ctx, nil)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, moving, test.ShouldBeFalse)
+		test.That(t, speed, test.ShouldEqual, 0.0)
+
+		// Change address in config copy to reference address of the remote2, make a copy of the config, and reconfigure
+		mainCfgCopy.Remotes[0].Address = addr2
+		mainCfgCopy2 := mainCfgCopy
+		mainRobot.Reconfigure(ctx, &mainCfgCopy)
```

Comment:
just to confirm, the intention here is to modify the already used `mainCfgCopy` and re-use for reconfiguration. `mainCfgCopy2` is meant for later in the flow and the only reason we make it now is because `Reconfigure` can modify the config, right?

Reply:
correct!

Reply:
cool... annoying that we have that situation. i wonder if our various public config-ingesting functions should make a copy before modifying the config. (thinking out loud, not something for this PR obviously)

Reply:
> public config-ingesting functions should make a copy before modifying the config

Yeah... particularly `Reconfigure`. I can file a ticket.

Reply:
https://viam.atlassian.net/browse/RSDK-8173